Add TypeScript definitions

jopemachine opened the pull request:
# Add typescript definitions

Thanks for the quick reply!

Here is the PR.

> Maybe do a pull request to add types to that package? Link to it here and I'll ping the maintainer to review it.

Ok, I will try to do a pull request to add types to cache-conf.

## Related

Fixes #127

* [cache-conf type definitions issue](https://github.com/SamVerschueren/cache-conf/issues/7)

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,393 @@
+	readonly variables?: Record<string, string>;
+}
+
+interface ActionElement {
```
You need to export all these types.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,393 @@
+
+export default interface Alfy {
+	/**
+	@param {Record<string, ScriptFilterItem>} items
```
Missing description. Doesn't need type annotation.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,393 @@
+	@param {string[]} list
+	@param {string|((item: string | Record<string, unknown>, input: string) => boolean)} item
+	@returns {string[]}
+	@description Same as matches(), but with alfy.input as input
```
```suggestion
	@description Same as matches(), but with `alfy.input` as input.
```

sindresorhus commented:
Make sure you follow https://github.com/sindresorhus/typescript-definition-style-guide

jopemachine commented on index.d.ts:
```diff
@@ -0,0 +1,393 @@
+	readonly url?: string;
+	readonly file?: string;
+	readonly auto?: string;
+}
```
I just added some comments about this.

I have never used `action` attributes, so I'm not sure if there my comments about this are all appropriate.

jopemachine commented:
I'm sorry for missing several conventions guides and thank you for pointing it out.

May I request code review again to find out if there are points that I'm still missing?

LitoMore commented on index.d.ts:
````diff
@@ -0,0 +1,460 @@
+	```
+	*/
+	userConfig: Map<string, string>;
+}
````
Newline required.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,460 @@
+export interface FetchOptions extends GotOptions<string> {
+	/**
+	Parse response body with JSON.parse and set accept header to application/json.
+	@default true
```
Too cramped.
```suggestion

	@default true
```

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,460 @@
+/**
+	Defines what to change when the modifier key is pressed.
+	When you release the modifier key, it returns to the original item.
+	@interface ModifierKeyItem
```
This is moot.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,460 @@
+}
+
+/**
+	Defines what to change when the modifier key is pressed.
```
Should not have indentation.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,460 @@
+	"action": ["Alfred is Great", "I use him all day long"]
+
+	For control over the content type of the action, you can use an object with typed keys as follows:
+	"action": {
```
Use a proper code block and `@example` tag. See the TS style guide.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,460 @@
+export default interface Alfy {
+	/**
+	Return output to Alfred.
+	@param items
```
Missing description.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,460 @@
+		themeBackground: string;
+		themeSelectionBackground: string;
+		themeSubtext: string;
+		data: string;
```
These properties need doc comments.

sindresorhus commented on index.d.ts:
````diff
@@ -0,0 +1,460 @@
+	```
+	*/
+	icon: {
+		get: (icon: string) => string;
````
Need doc comment.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,460 @@
+	Exports a Map with the user workflow configuration.
+	A workflow configuration allows your users to provide configuration information for the workflow. 
+	For instance, if you are developing a GitHub workflow, you could let your users provide their own API tokens.
+	See alfred-config for more details.
```
Should be linkified like the readme.

Make sure that all the text you copied from the readme retains its style, links, and inline code blocks.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,460 @@
+	/**
+	Whether the user currently has the workflow debugger open.
+	*/
+	debug: boolean
```
```suggestion
	debug: boolean;
```

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,460 @@
+	@param input
+	@param list
+	@param target
+	@returns string[] of items in list that case-insensitively contains input.
```
Don't mention the type. It's already clear from the actual types.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,460 @@
+	If you would like Alfred to skip this check as you are certain that the files you are returning exist, you can use "type": "file:skipcheck".
+	@default "default"
+	*/
+	readonly type?: "default" | "file" | "file:skipcheck";
```
Use single quotes everywhere.

LitoMore commented:
@jopemachine Would you mind upgrading xo to `0.39.1` in the pull request? The linter will help you fix all linter-related issues automatically.

LitoMore commented:
@jopemachine You could go through the [Testing](https://github.com/sindresorhus/typescript-definition-style-guide#testing) to learn how to use [tsd](https://github.com/SamVerschueren/tsd) to write tests for your TypeScript type definitions.

jopemachine commented:
@LitoMore, @sindresorhus I appreciate for detailed code reviews :)

I upgraded xo to `0.39.1` and fix some linting bugs (New bugs after upgrading `xo`)

And updated type declaration, added `tsd` test.

In tsd test, to test functions with return type void, I added `/* eslint-disable @typescript-eslint/no-confusing-void-expression */` comment on top.

If we don't need to test void returning functions, just let me know, I will remove the comment and tests.

jopemachine commented:
It seems that [update-notifier](https://github.com/yeoman/update-notifier) in xo `0.39.1` uses `node >= 10`,

So test fails in Node 8.

May I ask how I should handle this?

LitoMore commented:
> Test fails in Node 8.

@jopemachine Feel free to drop Node.js 8 from GitHub actions.

It seems the [got@9](https://github.com/sindresorhus/alfy/blob/main/.github/workflows/main.yml#L13) blocked us using Node.js 10 or higher. We could solve this problem later (in another PR) if you are interested.

jopemachine commented:
> > Test fails in Node 8.
> 
> @jopemachine Feel free to drop Node.js 8 from GitHub Actions.
> 
> It seems the [got@9](https://github.com/sindresorhus/alfy/blob/main/.github/workflows/main.yml#L13) blocked us using Node.js 10 or higher. We could solve this problem later (in another PR) if you are interested.

Of course, I'm happy to have the opportunity to contribute to open source.

It seems that `got@12.x` is ESM module, should we upgrade got to `got@11.8.2`?

LitoMore commented on package.json:
```diff
@@ -10,13 +10,14 @@
 	},
 	"engines": {
-		"node": ">=8"
+		"node": ">=10.x"
```
```suggestion
		"node": ">=10"
```

LitoMore commented:
>It seems that `got@12` is ESM, should we upgrade got to `got@11.8.2`?

Yes. Let's discuss this in the future pull request.

jopemachine commented:
> > It seems that `got@12` is ESM, should we upgrade got to `got@11.8.2`?
> 
> @jopemachine Yes. Let's discuss this in the future pull request.

@LitoMore I made a PR about this issue [here](https://github.com/sindresorhus/alfy/pull/140).

LitoMore commented on index.test-d.ts:
```diff
@@ -0,0 +1,34 @@
+
+expectType<void>(alfy.log('some message'));
+
+expectType<Promise<any>>(alfy.fetch('https://api.foo.com', {
```
Let's use a nonexistent URL instead.

```suggestion
expectType<Promise<any>>(alfy.fetch('https://foo.bar', {
```

LitoMore commented on package.json:
```diff
@@ -49,10 +50,13 @@
 		"read-pkg-up": "^6.0.0"
 	},
 	"devDependencies": {
+		"@types/got": "^9.6.12",
```
Got is written by TypeScript. Declarations are [already inside](https://github.com/sindresorhus/got/blob/f896aa52abc41fe40d4942da94a0408477358f14/package.json#L96).

LitoMore replied:
Now this can be removed. #140 merged.

LitoMore commented on package.json:
```diff
@@ -43,7 +44,7 @@
 		"esm": "^3.2.18",
 		"execa": "^2.0.4",
-		"got": "^9.3.2",
+		"got": "^11.8.2",
```
These changes should not appear in this PR. Could you rebase the `main` branch?

`git rebase -i main`. Then some conflicts need to be resolved when rebasing.

jopemachine replied:
If I remove `c0f6e87`, I think I also have to remove the 4 commits below to pass the tests (ava && tsd).

May I ask it is alright to remove these 5 commits?

LitoMore replied:
```shell
git checkout main
git pull upstream main
git checkout Add-types
git rebase -i main
git push origin Add-types -f
```

jopemachine replied:
Oh, I got it, thanks for guiding me :)
